fairscale/optim/oss.py

```diff
@@ -334,12 +321,110 @@ def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
             # Acknowledge broadcasts, and send this rank's shard when needed
             self._broadcast_state_dict()
 
+    def state_dict(self) -> Dict[str, Any]:
+        """Return the last known global optimizer state. The returned state is compatible with Pytorch, in that the
+        sharded properties are not exposed. It contains two entries:
+
+        * state - a dict holding current optimization state. Its content
+            differs between optimizer classes.
+
+        * param_groups - a dict containing all parameter groups
+
+        .. warning:
+            If the state has not been consolidated, this returns a shard's worth, not the global state.
+
+        .. warning:
+            Returning the global state is limited to the replica which was responsible for the consolidation.
+            The state may also not be up to date, depending on when `consolidate_state_dict` was last called.
+        """
+
+        if len(self._all_states) == 0:
+            logging.warning("Optimizer state has not been consolidated. Returning the local state")
+            logging.warning("Please call `consolidate_state_dict()` beforehand if you meant to save the global state")
+            return self.optim.state_dict()
+
+        # Unify the shard states and the state that pytorch would expect, given the model.
+        # Indexation needs several redirections, since each shard only knows a limited scope of the model
+        # - get the pytorch compliant parameter indexing
+        state_dict = super().state_dict()
+
+        # - get an id map which links the parameter id to the index in the reference state
+        global_id_map = [{id(p): i for pg in self.param_groups for i, p in enumerate(pg["params"])}]
```

Comment:
This looks wrong. You are creating a single dictionary wrapped in a list. Please add a test for the case where there are multiple param groups. I think it should be:
```suggestion
        global_id_map = [{id(p): i  for i, p in enumerate(pg["params"])} for pg in self.param_groups]
```

Reply:
makes sense, I had a quick thought about that but did not follow through. I'll fix and add a test